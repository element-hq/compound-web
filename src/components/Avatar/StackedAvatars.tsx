/*
Copyright 2023  New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import classnames from "classnames";
import React, { cloneElement } from "react";
import styles from "./Avatar.module.css";
import { Avatar } from "./Avatar";

import AvatarClipPath from "./avatar-clip.mask.svg";
import { createPortal } from "react-dom";

type StackedAvatarsProps = {
  className?: string;
  size: React.ComponentProps<typeof Avatar>["size"];
  avatars: Omit<React.ComponentProps<typeof Avatar>[], "type" | "size">;
};

// const AVATAR_MASK_ID = "cpdAvatarClipSvg";

/**
 * Renders a stack of avatars and clips the content appropriately.
 *
 * The `type` and `size` property of the children will be overriden so they
 * are all round and have the same size.
 */
export const StackedAvatars = ({
  children,
  size,
  className,
}: React.PropsWithChildren<StackedAvatarsProps>): React.JSX.Element => {
  return (
    <div className={classnames(styles["stacked-avatars"], className)}>
      {React.Children.map(children, (child) =>
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        cloneElement(child as any, {
          type: "round", // Only supports `round` avatars
          size, // Forces all avatars to be of the same size
        })
      )}
    </div>
  );
};
