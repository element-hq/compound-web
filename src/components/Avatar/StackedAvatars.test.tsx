/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { render } from "@testing-library/react";
import React from "react";

import { StackedAvatars } from "./StackedAvatars";

const originalImage = global.Image;

describe("StackedAvatars", () => {
  beforeEach(() => {
    global.Image = class extends originalImage {
      constructor() {
        super();
        setTimeout(() => {
          this.onload?.(new Event("load"));
        }, 50);
      }
    };
  });

  afterEach(() => {
    global.Image = originalImage;
  });

  const avatars = [
    { id: "@alice:example.org", name: "Alice" },
    { id: "@bob:example.org", name: "Bob" },
  ];

  it("renders", () => {
    const { asFragment } = render(
      <StackedAvatars avatars={avatars} size="32px" />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("adds the mask to the body", () => {
    render(<StackedAvatars avatars={avatars} size="32px" />);

    // We can't run better assertions as `SVG` loading is mocked out in jest
    expect(document.querySelectorAll("svg")).toHaveLength(1);

    render(<StackedAvatars avatars={avatars} size="32px" />);

    // We only one instance of the mask, ever!
    expect(document.querySelectorAll("svg")).toHaveLength(1);
  });
});
