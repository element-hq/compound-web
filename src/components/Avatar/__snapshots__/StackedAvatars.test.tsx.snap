// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StackedAvatars renders 1`] = `
<DocumentFragment>
  <div
    class="stacked-avatars"
  />
</DocumentFragment>
`;
